### ISSUE TITLE
Separate instructions for Debian

### ISSUE THREAD

**Issue by ivanistheone** [learningequality/kolibri-docs#60]
Installing from .deb from the PPA on Debian requires roughly the following instructions on Debian 9:

```
sudo apt-get install software-properties-common dirmngr
sudo su -c 'echo "deb http://ppa.launchpad.net/learningequality/kolibri/ubuntu cosmic main" > /etc/apt/sources.list.d/learningequality-ubuntu-kolibri-cosmic.list'
sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys DC5BAA93F9E4AE4F0411F97C74F88ADB3194DD81
sudo apt-get update
sudo apt-get install kolibri
```

Since instructions are sufficiently different from the default Ubuntu install steps, it might be worth having the separate section for Debain install instructions to guarantee optimal copy-and-pastability for successful install.

**Comment by radinamatic** [comment]
Sure thing @ivanistheone, will add it to the Linux installation chapter! :+1:

**Comment by jredrejo** [comment]
@ivanistheone @radinamatic These are alternative instructions
1. download Kolibri deb package
2. In the folder where kolibri has been downloaded `sudo dpkg -i kolibri-*deb`

Both are valid, but I am not sure which one is better for the user. @ivanistheone  steps would allow an easier upgrade path later just doing `apt-get update & apt-get install kolibri` but require more steps. I leave on you to decide.

**Comment by benjaoming** [comment]
For a quick fix, let's rename the sections:

"Debian / Ubuntu" => "Ubuntu"
"Raspberry Pi" => "Raspberry Pi / Debian"

**Comment by benjaoming** [comment]
@jredrejo is this input based on our current documentation? This method is already documented, even the discussion of pros/cons are already there:

![image](https://user-images.githubusercontent.com/374612/51469771-6802c500-1d72-11e9-907f-4999affc3925.png)

**Comment by jredrejo** [comment]
@benjaoming no, sorry. I am afraid I answered only based on the feedback Ivan provided. I didn't look at the current docs.

**Comment by radinamatic** [comment]
@ivanistheone @benjaoming @jredrejo Is this still extant, or can we consider it covered with the current docs?

**Comment by benjaoming** [comment]
@radinamatic you will find that the former PR was merged in the 0.11 release branch which unfortunately hasn't been sync'ed forwards :/

**Comment by jredrejo** [comment]
https://kolibri.readthedocs.io/en/latest/install.html#debian-ubuntu is still a mess. As @benjaoming points, the latest docs have not been synced.

https://github.com/learningequality/kolibri-docs/blob/develop/docs/install/_linux.rst is fine. I would remove `python-software-properties` because it's a dependency of `software-properties-common` so we don't need to specify it.

https://github.com/learningequality/kolibri-docs/blob/develop/docs/install/_rpi.rst I would remove all the install instructions and put a link to the Ubuntu/Debian steps. RPi uses Debian, same instructions apply. The rest of the info in this section is fine as they are good advices for RPi owners,

@benjaoming @radinamatic I don't have any problem on doing these changes myself. I just want to know if we all agree before. Also, I don't know (probably I don't have permissions) how to deploy the new docs to kolibri.readthedocs.io.

**Comment by radinamatic** [comment]
@jredrejo Thanks for reviewing these! :+1: 

Don't worry about changes, I'll incorporate the necessary in my current PR, according to your suggestions, after we merge in the changes from the 0.11 branch

**Comment by benjaoming** [comment]
@jredrejo it's worth waiting for the 0.11.x to be sync'ed into 0.12.x

> I would remove python-software-properties

I don't want to do archeology on this, but there was a reason why it was there, I suspect some old versions needed it. I'd just keep it since it's harmless and only for PPA purposes.

**Comment by benjaoming** [comment]
@jredrejo re: RPi installation

> I would remove all the install instructions and put a link to the Ubuntu/Debian steps.

I'd like to keep them RPi-specific :) As you can see from the first steps, they address the `cffi` library. Including those commands in the other steps would make those steps more messy.

I contemplate moving the Raspbian install instructions to the tutorial, such that people won't miss the context, see the current reference given here: https://kolibri.readthedocs.io/en/release-v0.12.x/install/tutorials/rpi.html#installing-kolibri

**Comment by jredrejo** [comment]
> @jredrejo it's worth waiting for the 0.11.x to be sync'ed into 0.12.x
> 
> > I would remove python-software-properties
> 
> I don't want to do archeology on this, but there was a reason why it was there, I suspect some old versions needed it. I'd just keep it since it's harmless and only for PPA purposes.

I have done the archelogy:
https://packages.ubuntu.com/xenial/software-properties-common
you can see that ubuntu 16.04 includes it as a depedency. Our docs say that the instructions are for >=16.04.
I don't know if older versions of Ubuntu didn't have the dependency. Probably they haven't and that was the reason to include it in the docs, but the versions we are supporting don't need it.

**Comment by benjaoming** [comment]
@jredrejo Oh `python-software-properties` is actually already removed. It's because you were looking in the `develop` which isn't updated... changes in 0.11.x haven't traversed yet. I'm working on that

**Comment by jredrejo** [comment]
> @jredrejo re: RPi installation
> 
> > I would remove all the install instructions and put a link to the Ubuntu/Debian steps.
> 
> I'd like to keep them RPi-specific :) As you can see from the first steps, they address the `cffi` library. Including those commands in the other steps would make those steps more messy.
> 
> I contemplate moving the Raspbian install instructions to the tutorial, such that people won't miss the context, see the current reference given here: https://kolibri.readthedocs.io/en/release-v0.12.x/install/tutorials/rpi.html#installing-kolibri

Honestly, I don't understand it. Raspbian is a Debian system. I have had a RPi working following the Debian instructions for months without problems . I don't understand why the instructions say to upgrade cffi if the kolibri package brings its own version and ignores the one the system uses.

About https://kolibri.readthedocs.io/en/release-v0.12.x/install/tutorials/rpi.html#installing-kolibri , I like most of that doc ( I would include `kolibri-server` instead of the nginx part, and would ensure the docs are updated, as Raspbian has been updated). RPi has some special characteristics as the firmware or the hardware setup and I agree it deserves a special page. However, I think that all the things that are common to any device should be in the common page. The hotspot section is an example of it. I really like the instructions written in that page, but those same instructions apply to any Debian/Ubuntu derivative (as far as they have a wifi card/dongle), they are not Raspberry specific.

**Comment by benjaoming** [comment]
> I would include kolibri-server instead of the nginx part, and would ensure the docs are updated, as Raspbian has been updated

The whole thing was written before `kolibri-server` came into existence, so let's get it updated once relevant.

Note that it takes a lot of time to sit down and create and test these instructions.

> I don't understand why the instructions say to upgrade cffi if the kolibri package brings its own version and ignores the one the system uses.

You can try on a blank RPI without this instruction. It was added because it was needed.

**Comment by benjaoming** [comment]
@jredrejo I don't think the scope of this issue is to generalize the RPI tutorial to all Debian-based systems ;)

**Comment by jredrejo** [comment]
> 
> You can try on a blank RPI without this instruction. It was added because it was needed.

I am not saying it was not needed. I am saying it's not needed anymore. cffi is included in the kolibri package. Upgrading/installing it in the RPi does not produce any difference. I have tested it in a blank RPI and it works.

I don't think it's good to release obsolete instructions after a revision of the documentation.

> @jredrejo I don't think the scope of this issue is to generalize the RPI tutorial to all Debian-based systems ;)

No, it's not, but I thought that doing a proper organization of the contents is.

**Comment by benjaoming** [comment]
> I am not saying it was not needed. I am saying it's not needed anymore. cffi is included in the kolibri package. Upgrading/installing it in the RPi does not produce any difference. I have tested it in a blank RPI and it works.

Weird, I have missed out on something: Where is `cffi` included? I don't find it in Kolibri itself, nor as a dependency. Raspbian used to ship with an incompatible version. I think it broke with the pre-compiled bundled cryptography, not sure.

> I don't think it's good to release obsolete instructions after a revision of the documentation.

I think it depends on whether Jessie (Lite) is supposed to be unsupported.

I'm vary of regressions because all this testing and digging around, responding to user feedback etc. was quite costly. I think that the introduction of `cffi` in the intructions came at a time when Raspbian Stretch was already released (August 2017). You can see the versions provided here: https://packages.debian.org/search?keywords=python3-cffi

https://github.com/learningequality/kolibri-docs/issues/33#issuecomment-423743951

We haven't heard of any issues from Stretch users because of the `cffi` installation, but I guess we can move things into a box for Jessie-based RPis.

Did you do tests on Raspian Stretch Lite? The intention is to support this image as a baseline.

> No, it's not, but I thought that doing a proper organization of the contents is.

From a user's perspective, I think it's very nice to have a Raspberry Pi tutorial be about that particular hardware setup. A lot of our instructions relate to this device, with its onboard WIFI, links to downloading Raspbian etc.

As I mentioned before, I think it's better to include the Raspbian installation instructions in the Tutorial itself and let that be the main installation manual for RPi, putting an anchor-link in place of the current RPi install instructions in `_rpi.rst`

**Comment by jredrejo** [comment]
> Weird, I have missed out on something: Where is `cffi` included? I don't find it in Kolibri itself, nor as a dependency. Raspbian used to ship with an incompatible version. I think it broke with the pre-compiled bundled cryptography, not sure.

You can compare the output of
https://packages.debian.org/jessie/armel/python3-cffi/filelist
with the output of
`dpkg -L kolibri|grep armv|grep cffi`
on any system with the kolibri package installed

I don't know how it happened but we are clearly distributing it and kolibri is using it because I have checked kolibri is running in systems without the python3-cffi package installed (and without doing the `pip install cffi` command the current docs say)

**Comment by benjaoming** [comment]
@jredrejo thanks, that might explain it! `cffi` is bundled as a dependency of `cryptography` for the available archs on PyPi -- it isn't bundled for arm* archs. I think the issue would possibly have occurred due to the availability of `cryptography` somehow on the system I've been building these docs with? Maybe the availability of various archs has been flimpsy or I've somehow managed to run into this issue.

But `cryptography` itself would seem most critical on small devices like an RPi.

-----

I want to stick to an actionable conclusion. Does anyone have any objections about this?

> As I mentioned before, I think it's better to include the Raspbian installation instructions in the Tutorial itself and let that be the main installation manual for RPi, putting an anchor-link in place of the current RPi install instructions in _rpi.rst

**Comment by indirectlylit** [comment]
>  I think it's better to include the Raspbian installation instructions in the Tutorial itself and let that be the main installation manual for RPi, putting an anchor-link in place of the current RPi install instructions in _rpi.rst

Just to clarify, is this the proposal? (I don't quite understand the 'anchor-link' phrase)

![image](https://user-images.githubusercontent.com/2367265/57317943-188c9e80-70ae-11e9-97fb-b2b3fbe23f8c.png)


If so, that seems like a good plan to me

**Comment by benjaoming** [comment]
Yes @indirectlylit - read my mind :heart:

**Comment by radinamatic** [comment]
Glad we have a consensus, go team! :+1: 

I'll include the proposed change in my current PR. The only thing I'm unsure of is what to do with the ``cffi`` library, does it stay included as [in our current 0.12 branch](https://github.com/learningequality/kolibri-docs/blob/release-v0.12.x/docs/install/_rpi.rst)?
